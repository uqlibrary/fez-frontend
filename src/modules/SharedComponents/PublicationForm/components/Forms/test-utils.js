import { useForm } from 'react-hook-form';
import { render, WithReduxStore } from 'test-utils';
import React from 'react';
import { formValues } from '../PublicationForm';

export const setupWrapper = (Form, props) => {
    const Wrapper = () => {
        const {
            control,
            formState: { isSubmitting },
        } = useForm();

        // eslint-disable-next-line react/prop-types
        return <Form {...{ control, isSubmitting, ...props, formValues: formValues(props?.formValues || {}) }} />;
    };

    return render(
        <WithReduxStore>
            <Wrapper />
        </WithReduxStore>,
    );
};
