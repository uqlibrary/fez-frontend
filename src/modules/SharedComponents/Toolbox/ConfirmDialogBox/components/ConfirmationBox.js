import React from 'react';
import PropTypes from 'prop-types';
import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogActions from '@mui/material/DialogActions';
import DialogContentText from '@mui/material/DialogContentText';
import DialogContent from '@mui/material/DialogContent';
import Grid from '@mui/material/Grid';

export const ConfirmationBox = ({
    actionButtonColor,
    actionButtonVariant,
    cancelButtonColor,
    classes = {},
    confirmationBoxId,
    InputForm,
    hideActionButton = false,
    hideCancelButton = false,
    isOpen = false,
    locale = {
        confirmationTitle: 'Confirmation',
        confirmationMessage: 'Are you sure?',
        cancelButtonLabel: 'No',
        confirmButtonLabel: 'Yes',
        alternateActionButtonLabel: 'Cancel',
    },
    onAction,
    onAlternateAction,
    onCancelAction,
    onClose,
    showAlternateActionButton = false,
    showInputForm = false,
}) => {
    const _onAction = () => {
        onClose();
        onAction();
    };

    const _onCancelAction = () => {
        onClose();
        !!onCancelAction && onCancelAction();
    };

    const _onAlternateAction = () => {
        onClose();
        !!onAlternateAction && onAlternateAction();
    };

    return (
        <Dialog
            style={{ padding: 6 }}
            open={isOpen}
            id={confirmationBoxId}
            data-testid={confirmationBoxId}
            sx={{ ...classes }}
        >
            <DialogTitle data-testid="message-title">{locale.confirmationTitle}</DialogTitle>
            <DialogContent>
                <DialogContentText data-testid="message-content">{locale.confirmationMessage}</DialogContentText>
                {!!showInputForm && <InputForm />}
            </DialogContent>
            <DialogActions>
                <Grid container spacing={1}>
                    <Grid item xs sx={{ display: { xs: 'none', sm: 'block' } }} />

                    {!hideActionButton && (
                        <Grid item xs={12} sm={'auto'}>
                            <Button
                                {...(!!actionButtonVariant ? { variant: actionButtonVariant } : {})}
                                children={locale.confirmButtonLabel}
                                autoFocus
                                color={actionButtonColor || 'primary'}
                                fullWidth
                                onClick={_onAction}
                                id="confirm-action"
                                data-analyticsid={`confirm-${confirmationBoxId}`}
                                data-testid={`confirm-${confirmationBoxId}`}
                            />
                        </Grid>
                    )}
                    {showAlternateActionButton && (
                        // an optional middle button that will display in a warning colour
                        <Grid item xs={12} sm={'auto'}>
                            <Button
                                variant={'contained'}
                                sx={{
                                    color: 'white.main',
                                    backgroundColor: 'warning.main',
                                    '&:hover': {
                                        backgroundColor: 'warning.dark',
                                    },
                                }}
                                children={locale.alternateActionButtonLabel}
                                fullWidth
                                onClick={_onAlternateAction}
                                id="confirm-alternate-action"
                                data-analyticsid={`confirm-alternate-${confirmationBoxId}`}
                                data-testid={`confirm-alternate-${confirmationBoxId}`}
                            />
                        </Grid>
                    )}
                    {!hideCancelButton && (
                        <Grid item xs={12} sm={'auto'}>
                            <Button
                                variant={'contained'}
                                color={cancelButtonColor || 'primary'}
                                children={locale.cancelButtonLabel}
                                fullWidth
                                onClick={_onCancelAction}
                                id="confirm-cancel-action"
                                data-analyticsid={`cancel-${confirmationBoxId}`}
                                data-testid={`cancel-${confirmationBoxId}`}
                            />
                        </Grid>
                    )}
                </Grid>
            </DialogActions>
        </Dialog>
    );
};

ConfirmationBox.propTypes = {
    actionButtonColor: PropTypes.string,
    actionButtonVariant: PropTypes.string,
    cancelButtonColor: PropTypes.string,
    classes: PropTypes.object,
    confirmationBoxId: PropTypes.string.isRequired,
    hideActionButton: PropTypes.bool,
    hideCancelButton: PropTypes.bool,
    InputForm: PropTypes.func,
    isOpen: PropTypes.bool,
    locale: PropTypes.object,
    onAction: PropTypes.func,
    onCancelAction: PropTypes.func,
    onAlternateAction: PropTypes.func,
    onClose: PropTypes.func,
    showAlternateActionButton: PropTypes.bool,
    showInputForm: PropTypes.bool,
};

export default React.memo(ConfirmationBox);
