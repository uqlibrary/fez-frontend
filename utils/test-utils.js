/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';

import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';

import mediaQuery from 'css-mediaquery';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');
const mime = require('mime-types');

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@testing-library/dom';
import preview from 'jest-preview';

export const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <LocalizationProvider dateAdapter={AdapterMoment}>{props.children}</LocalizationProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => {
    return {
        user: userEvent.setup(),
        ...render(ui, { wrapper: AllTheProviders, ...options }),
    };
};

export const WithRouter = ({ children, route = '/', initialEntries = [route] }) => {
    const routes = [{ path: route, element: children }];
    const router = createMemoryRouter(routes, { initialEntries: initialEntries });
    return <RouterProvider router={router} />;
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

export const assertTooltipText = async (trigger, tooltipText) => {
    expect(trigger).toBeInTheDocument();
    await userEvent.hover(trigger);
    await waitFor(() => {
        expect(screen.getByRole('tooltip', { name: tooltipText, hidden: true })).toBeVisible();
    });
};

const extensionToMimeMap = {
    tiff: 'image/tiff',
    wav: 'audio/wave',
    zip: 'application/zip',
};
const mimeType = filename => extensionToMimeMap[filename.split('.').pop()];

export const defaultDatastream = {
    dsi_embargo_date: null,
    dsi_open_access: 1,
    dsi_label: '',
    dsi_copyright: null,
    dsi_state: 'A',
    dsi_size: 27932352,
};

export const createFezDatastreamInfoArray = (datastreams, pid = null, withPreview = true) => {
    let processed = datastreams.filter(datastream => typeof datastream === 'object' || typeof datastream === 'string');
    if (withPreview) {
        processed = [];
        datastreams.forEach(datastream => {
            processed.push(datastream);
            const filename = typeof datastream === 'object' ? datastream.dsi_dsid : datastream;
            const mimetype = mimeType(filename);
            // bail in case it's not derivable
            if (
                !filename.includes('.pdf') &&
                !mimetype.includes('image') &&
                !mimetype.includes('audio') &&
                !mimetype.includes('video')
            ) {
                return;
            }
            // eslint-disable-next-line no-nested-ternary
            const derivativeMimetype = mimetype.includes('audio')
                ? 'audio/mp3'
                : mimetype.includes('video')
                ? 'video/mp4'
                : 'image/jpg';

            (derivativeMimetype.includes('image') ? ['preview_', 'thumbnail_', 'web_'] : ['']).forEach(prefix => {
                processed.push({
                    dsi_dsid: `${prefix}${filename.split('.')[0]}_t.${derivativeMimetype.split('/').pop()}`,
                    dsi_mimetype: derivativeMimetype,
                });
            });
        });
    }

    return processed.map((datastream, index) => ({
        ...defaultDatastream,
        ...{
            dsi_id: index + 1,
            dsi_pid: pid,
            dsi_order: index + 1,
            dsi_mimetype: mime.lookup(typeof datastream === 'object' ? datastream.dsi_dsid : datastream),
        },
        ...(typeof datastream === 'object'
            ? datastream
            : {
                  dsi_dsid: datastream,
              }),
    }));
};

export const withDatastreams = (sources, datastreams, callback) =>
    datastreams.map(item => {
        const source = sources.find(datastream =>
            datastream.dsi_dsid.includes(
                item.dsi_dsid
                    .replace('_t.', '.')
                    .replace('preview_', '')
                    .replace('thumbnail_', '')
                    .replace('web_', '')
                    .split('.')[0],
            ),
        );
        const isDerivative = item.dsi_dsid !== source.dsi_dsid;
        callback(item, source, isDerivative);
    });

export const getDatastreamByFilename = (filename, datastreams) =>
    datastreams.find(datastream => datastream.dsi_dsid === filename);

export const createMatchMedia = width => {
    return query => ({
        matches: mediaQuery.match(query, { width }),
        /* istanbul ignore next */
        addListener: () => {},
        /* istanbul ignore next */
        removeListener: () => {},
    });
};

export const getFilenameExtension = filename => filename.split('.').pop();
export const getFilenameBasename = filename =>
    filename.replace(new RegExp(`/\.${getFilenameExtension(filename)}$/`), '');
export const addFilesToFileUploader = files => {
    const { screen, fireEvent } = reactTestingLib;
    // create a list of Files
    const fileList = files.map(file => {
        // if file it's a string, treat it as a filename
        if (typeof file === 'string') {
            return new File([getFilenameBasename(file)], file, { type: `image/${getFilenameExtension(file)}` });
        }
        // otherwise expect it to be a object with filename and mimeType keys
        return new File([getFilenameBasename(file.filename)], file.filename, { type: file.mimeType });
    });
    // drag and drop files
    fireEvent.drop(screen.getByTestId('fez-datastream-info-input'), {
        dataTransfer: {
            files: fileList,
            types: ['Files'],
        },
    });
};
export const setFileUploaderFilesToClosedAccess = async (files, timeout = 500) => {
    const { fireEvent } = reactTestingLib;
    // set all files to closed access
    for (const file of files) {
        const index = files.indexOf(file);
        await waitFor(() => screen.getByText(new RegExp(getFilenameBasename(file))), { timeout });
        fireEvent.mouseDown(screen.getByTestId(`dsi-open-access-${index}-select`));
        fireEvent.click(screen.getByRole('option', { name: 'Closed Access' }));
    }
};

export const assertEnabled = (resolver, ...items) => items.forEach(item => expect(resolver(item)).toBeEnabled());
export const assertDisabled = (resolver, ...items) => items.forEach(item => expect(resolver(item)).toBeDisabled());

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    withRedux,
    AllTheProviders,
    WithReduxStore,
    assertTooltipText,
    WithRouter,
    createFezDatastreamInfoArray,
    getDatastreamByFilename,
    withDatastreams,
    createMatchMedia,
    preview,
    userEvent,
    assertEnabled,
    assertDisabled,
    getFilenameExtension,
    getFilenameBasename,
    addFilesToFileUploader,
    setFileUploaderFilesToClosedAccess,
};
