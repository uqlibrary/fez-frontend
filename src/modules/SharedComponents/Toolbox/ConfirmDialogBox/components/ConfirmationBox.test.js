import React from 'react';
import ConfirmationBox from './ConfirmationBox';
import { rtlRender, fireEvent } from 'test-utils';

function setup(testProps = {}, renderer = rtlRender) {
    const props = {
        hideCancelButton: false,
        isOpen: false,
        locale: {
            confirmationTitle: 'Confirmation',
            confirmationMessage: 'Are you sure?',
            cancelButtonLabel: 'No',
            confirmButtonLabel: 'Yes',
            alternateActionButtonLabel: 'Cancel',
        },
        onAction: jest.fn(),
        onAlternateAction: jest.fn(),
        onCancelAction: jest.fn(),
        onClose: jest.fn(),
        showAlternateActionButton: false,
        confirmationBoxId: 'confirmation-box',
        ...testProps,
    };

    return renderer(<ConfirmationBox {...props} />);
}

describe('ConfirmationBox component', () => {
    it('should render confirmation box with default props', () => {
        const { container } = rtlRender(<ConfirmationBox confirmationBoxId="test" />);
        expect(container).toMatchSnapshot();
    });

    it('should render confirmation box', () => {
        const { getByTestId, getByText } = setup({ isOpen: true });
        expect(getByText('Are you sure?')).toBeInTheDocument();
        expect(getByTestId('confirm-confirmation-box')).toBeInTheDocument();
        expect(getByTestId('cancel-confirmation-box')).toBeInTheDocument();
    });

    it('should render alternate action button', () => {
        const { getByTestId, getByText } = setup({ isOpen: true, showAlternateActionButton: true });
        expect(getByText('Are you sure?')).toBeInTheDocument();
        expect(getByTestId('confirm-confirmation-box')).toBeInTheDocument();
        expect(getByTestId('cancel-confirmation-box')).toBeInTheDocument();
        expect(getByTestId('confirm-alternate-confirmation-box')).toBeInTheDocument();
    });

    it('should call confirm action', () => {
        const onActionFn = jest.fn();
        const onCloseFn = jest.fn();
        const { getByTestId } = setup({ isOpen: true, onAction: onActionFn, onClose: onCloseFn });
        fireEvent.click(getByTestId('confirm-confirmation-box'));
        expect(onActionFn).toBeCalled();
        expect(onCloseFn).toBeCalled();
    });

    it('should call cancel action', () => {
        const onCancelActionFn = jest.fn();
        const onCloseFn = jest.fn();
        const { getByTestId } = setup({ isOpen: true, onCancelAction: onCancelActionFn, onClose: onCloseFn });
        fireEvent.click(getByTestId('cancel-confirmation-box'));
        expect(onCancelActionFn).toBeCalled();
        expect(onCloseFn).toBeCalled();
    });

    it('should call alternate action', () => {
        const onAlternateActionFn = jest.fn();
        const onCloseFn = jest.fn();
        const { getByTestId } = setup({
            isOpen: true,
            onAlternateAction: onAlternateActionFn,
            onClose: onCloseFn,
            showAlternateActionButton: true,
        });
        fireEvent.click(getByTestId('confirm-alternate-confirmation-box'));
        expect(onAlternateActionFn).toBeCalled();
        expect(onCloseFn).toBeCalled();
    });
});
