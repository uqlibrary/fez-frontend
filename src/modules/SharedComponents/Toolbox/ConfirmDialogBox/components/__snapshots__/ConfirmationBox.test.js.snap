// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationBox component should render confirmation box with default props 1`] = `<div />`;
